presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewInformationExtractor.java

```diff
@@ -173,5 +168,26 @@ public boolean isDistinct()
         {
             return isDistinct;
         }
+
+        public Optional<Set<Expression>> getGroupBy()
+        {
+            return groupBy;
+        }
+
+        public boolean presentInGroupBy(GroupingElement groupingElement)
+        {
+            for (Expression expression : groupingElement.getExpressions()) {
+                if (!presentInGroupBy(expression)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        public boolean presentInGroupBy(Expression expression)
+        {
+            return groupBy.isPresent() &&
+                    (groupBy.get().contains(expression) || !baseToViewColumnMap.containsKey(expression));
```

Comment:
I find this check very confusing
"baseToViewColumnMap.containsKey(expression)"

I think we should not mix column mapping containment with group by clause. I think this is adding false positive, can you share an example why this is required?
Even if it is required, we should let the consumer of utility mix this logic. Group by should only worry about group by fields.

Reply:
Materialized View definition:
```SQL
CREATE MATERIALIZED VIEW mv
SELECT SUM(a) AS a, b FROM base GROUP BY c
```

The following query cannot be optimized, as `c` is not a selected column of the MV, even it is presents in the `GROUP BY`.
```SQL
SELECT SUM(a) FROM base GROUP BY c
```

If it passed this check, the query would be rewritten into
```SQL
SELECT SUM(a) FROM mv GROUP BY c
```
which is incorrect.

I agree with you checking the column containment in a method named `presentInGroupBy()` is confusing. How about the new name `validateExpressionForGroupBy()`?

Reply:
I think one of the main issue here is that information extractor is also doing validation as well. We should probably extract this logic into some utility.